lib/src/components/NxList/NxList.tsx

```diff
@@ -0,0 +1,120 @@
+import { NxListComponent, nxListPropTypes } from './types';
+
+const mutationObserverConfig = { subtree: false, childList: true, attributes: false, characterData: false };
+
+const NxList = forwardRef<HTMLUListElement, NxListProps>((props: NxListProps, ref) => {
+  const {className, children, bulleted, emptyMessage, isLoading = false, error, retryHandler, ...attrs } = props;
+  const classNames = classnames(className, 'nx-list', {'nx-list--bulleted': bulleted});
+  const [title, otherChildren] = splitOutFirst(NxH3, children);
+  const [isEmpty, setIsEmpty] = useState(false);
+  const ulRef = useRef<HTMLUListElement | null>(null);
+  const emptyListRef = useRef<HTMLLIElement>(null);
+
+  const updateIsEmpty = useCallback(function updateIsEmpty() {
+    if (ulRef.current) {
+      const listItems = ulRef.current.children;
+
+      //If there are no children, simply set isEmpty to true and return.
+      if (listItems.length === 0) {
+        setIsEmpty(true);
+        return;
+      }
+
+      //NxH3 is treated as a 'non-child' as we still would want to show a title for an empty list.
+      if (!title) {
+        setIsEmpty(!listItems.length || (listItems.length === 1 && listItems.item(0) === emptyListRef.current));
+      }
+      else {
+        setIsEmpty((listItems.length === 1 && listItems.item(0)?.tagName === 'H3')
+          || (listItems.length === 2 && (listItems.item(0)?.tagName === 'H3'
+              && listItems.item(1) === emptyListRef.current))
+        );
+      }
+    }
+  }, []);
+
+  useEffect(updateIsEmpty, []);
+  useMutationObserver(ulRef, updateIsEmpty, mutationObserverConfig);
+
+  if (isEmpty && !isLoading && !error) {
+    if (!emptyMessage) {
+      console.warn('emptyMessage is required when no list items are to be rendered');
+    }
+  }
+
+  const nxListEmpty = (
+    <li ref={emptyListRef} className="nx-list__item nx-list__item--empty">
+      <span className="nx-list__text">
+        {emptyMessage ? emptyMessage : 'This list is empty.'}
+      </span>
+    </li>
+  );
+
+  const nxListLoading = (
+    <li className='nx-list__item'>
+      <NxLoadingSpinner />
+    </li>
+  );
+
+  const nxListError = (
+    <li className='nx-list__item nx-list__item--error'>
+      <NxLoadError error={error} retryHandler={retryHandler} />
+    </li>
+  );
+
+  // Using ref.current in forwardRef see:
+  // https://stackoverflow.com/questions/62238716/using-ref-current-in-react-forwardref
+  return (
+    <ul ref={(node) => {
+      ulRef.current = node;
+      if (typeof ref === 'function') {
+        ref(node);
+      }
+      else if (ref) {
+        (ref as MutableRefObject<HTMLUListElement | null>).current = node;
+      }
+    }}
+        className={classNames}
+        {...attrs}
+    >
+      {title}
+      {isLoading && nxListLoading}
+      {!!error && !isLoading && nxListError}
+      {!isLoading && !error && otherChildren}
+      {!(isLoading || error) && isEmpty && nxListEmpty}
+    </ul>
+  );
+}) as NxListComponent;
```

Comment:
I don't think this cast, or even the existence of this explicit type, is necessary.  `NxTable` does fine without something like this.

Reply:
> I don't think this cast, or even the existence of this explicit type, is necessary. `NxTable` does fine without something like this.

I tried messing around with this, but it seems that `NxList` itself is a _forwarded_ component in a sense that it has refs on it using `forwardRef`. `NxTable`, on the other hand, is simply defined a function. I wasn't able to get by TypeScript errors without using the `NxListComponent` type. From what I noticed, `forwardRefs`, TypeScript and dot notations do not seem to play well together without explicitly defining types.

Reply:
> I tried messing around with this, but it seems that NxList itself is a forwarded component in a sense that it has refs on it using forwardRef. NxTable, on the other hand, is simply defined a function. I wasn't able to get by TypeScript errors without using the NxListComponent type. From what I noticed, forwardRefs, TypeScript and dot notations do not seem to play well together without explicitly defining types.

Right, the trick is to use `Object.assign` to define the entire `NxList` object in one statement.  We do this in `SimpleComponents.tsx` for the same reason.  So it would be like
```
const NxList = Object.assign(forwardRef(...), { Item: NxListItem, ... });
```